spirv_msl.cpp

```diff
@@ -19397,6 +19421,68 @@ void CompilerMSL::emit_mesh_tasks(SPIRBlock &block)
 	statement("return;");
 }
 
+void CompilerMSL::emit_workgroup_initialization(const SPIRVariable &var)
+{
+	auto &type = get_variable_data_type(var);
+
+	begin_scope();
+
+	if (type.array.empty() && type.member_types.empty())
+	{
+		// For simple shared variables, we just initialize it in thread 0 of the block
+		// We use short to represent bool for threadgroup variable to workaround compiler bug,
+		// so we do a temporary fixup here. Alas. (see the type_to_glsl method)
+		bool is_boolean = type.basetype == SPIRType::Boolean;
+		if (is_boolean) {
+			type.basetype = SPIRType::Short;
+		}
+		statement("if (gl_LocalInvocationIndex == 0)");
+		begin_scope();
+		statement(join(to_name(var.self), " = ", to_initializer_expression(var), ";"));
+		end_scope();
+		if (is_boolean) {
+			type.basetype = SPIRType::Boolean;
+		}
+	} else {
+		// Otherwise, we use a loop to cooperatively initialize the memory within the group
+
+		// First, we define a few variable names;
+		string var_name = to_name(var.self);
+		string var_ptr_name = join(var_name, "_ptr");
+		string var_size_name = join(var_name, "_sz");
+		string var_pos_name = join(var_name, "_pos");
+		string var_stride_name = join(var_name, "_stride");
+		string var_ptr2_name = join(var_name, "_ptr2");
+
+		statement("threadgroup uint *", var_ptr_name, " = (threadgroup uint *)&", var_name, ";");
```

Comment:
Gross, but it's certainly not worse than the alternative ...

Reply:
This is usually done at IR-level transformation rather than in the source. As I couldn't find a reliable way to let apple's compiler initialize threadgroup variables, I have to do it myself. My code is inspired by the NIR lowering pass in mesa. Thankfully msl is flexible enough to allow this in a straightforward manner.